palace/fem/fespace.hpp

```diff
@@ -35,6 +38,167 @@ class FiniteElementSpace : public mfem::ParFiniteElementSpace
   std::size_t GetId() const;
 };
 
+//
+// An AuxiliaryFiniteElement space is just a FiniteElementSpace which allows for lazy
+// construction of the interpolation operator (discrete gradient or curl) from the primal
+// space to this one.
+//
+class AuxiliaryFiniteElementSpace : public FiniteElementSpace
+{
+private:
+  const FiniteElementSpace &primal_fespace;
+  mutable std::unique_ptr<Operator> G;
+
+  const Operator &BuildDiscreteInterpolator() const;
+
+public:
+  template <typename... T>
+  AuxiliaryFiniteElementSpace(const FiniteElementSpace &primal_fespace, T &&...args)
+    : FiniteElementSpace(std::forward<T>(args)...), primal_fespace(primal_fespace)
+  {
+  }
+
+  // Return the discrete gradient or discrete curl matrix interpolating from the auxiliary
+  // to the primal space, constructing it on the fly as necessary.
+  const Operator &GetDiscreteInterpolator() const
+  {
+    return G ? *G : BuildDiscreteInterpolator();
+  }
+};
+
+//
+// A collection of FiniteElementSpace objects constructed on the same mesh with the ability
+// to construct the prolongation operators between them as needed.
+//
+class FiniteElementSpaceHierarchy
+{
+protected:
+  std::vector<std::unique_ptr<FiniteElementSpace>> fespaces;
+  mutable std::vector<std::unique_ptr<Operator>> P;
+
+  const Operator &BuildProlongationAtLevel(std::size_t l) const;
+
+public:
+  FiniteElementSpaceHierarchy() = default;
+  FiniteElementSpaceHierarchy(std::unique_ptr<FiniteElementSpace> &&fespace)
+  {
+    AddLevel(std::move(fespace));
+  }
+
+  auto GetNumLevels() const { return fespaces.size(); }
+  auto size() const { return GetNumLevels(); }
+  bool empty() const { return GetNumLevels() == 0; }
```

Comment:
I get the consistency with stl, but for consistency with all other methods and to avoid inconsistency if other "stl-like" methods appear, probably worth going with Camel. 

The only exception i can think of for this is `begin()` and `end()` to work with range loops, which actually should probably be added, given this is basically a thin wrapper for a std::vector. I would probably go for it referring to the fespaces, but could refer to a pair of space and prolongation if wanted. This would be a bit of a deviation from mfem, but would allow for range loops in a few places elsewhere.

Reply:
Sounds good, fixed.